cache/remotecache/ipfs/ipfs.go

```diff
@@ -0,0 +1,607 @@
+		if err := sh.FilesRm(ctx, ce.config.MFSDir, true); err != nil {
+			return nil, err
+		}
+	}
+
+	logrus.Debugf("creating path %s", ce.config.MFSDir)
+	if err := sh.FilesMkdir(ctx, ce.config.MFSDir); err != nil {
+		return nil, err
+	}
+
+	config, descs, err := ce.chains.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, l := range config.Layers {
+		dgstPair, ok := descs[l.Blob]
+		if !ok {
+			return nil, errors.Errorf("missing blob %s", l.Blob)
+		}
+		if dgstPair.Descriptor.Annotations == nil {
+			return nil, errors.Errorf("invalid descriptor without annotations")
+		}
+		var diffID digest.Digest
+		v, ok := dgstPair.Descriptor.Annotations["containerd.io/uncompressed"]
+		if !ok {
+			return nil, errors.Errorf("invalid descriptor without uncompressed annotation")
+		}
+		dgst, err := digest.Parse(v)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse uncompressed annotation")
+		}
+		diffID = dgst
+
+		path := blobKey(ce.config, dgstPair.Descriptor.Digest.String())
+		exists, err := blobExists(ctx, sh, path)
+		if err != nil {
+			return nil, err
+		}
+
+		logrus.Debugf("layers %s exists = %t", path, exists)
+
+		if !exists {
+			layerDone := progress.OneOff(ctx, fmt.Sprintf("writing layer %s", l.Blob))
+			dt, err := content.ReadBlob(ctx, dgstPair.Provider, dgstPair.Descriptor)
+			if err != nil {
+				return nil, layerDone(err)
+			}
+
+			if err := writeToTmp(filepath.Join(tmpDir, path), dt); err != nil {
+				return nil, layerDone(err)
+			}
+
+			_ = layerDone(nil)
+			logrus.Infof("wrote layer %s", l.Blob)
+		}
+
+		la := &v1.LayerAnnotations{
+			DiffID:    diffID,
+			Size:      dgstPair.Descriptor.Size,
+			MediaType: dgstPair.Descriptor.MediaType,
+		}
+		if v, ok := dgstPair.Descriptor.Annotations["buildkit/createdat"]; ok {
+			var t time.Time
+			if err := (&t).UnmarshalText([]byte(v)); err != nil {
+				return nil, err
+			}
+			la.CreatedAt = t.UTC()
+		}
+		config.Layers[i].Annotations = la
+	}
+
+	dt, err := json.Marshal(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal config")
+	}
+
+	manifestFilePath := manifestKey(ce.config, "manifest")
+	if err := writeToTmp(filepath.Join(tmpDir, manifestFilePath), dt); err != nil {
+		return nil, errors.Wrapf(err, "error writing manifest")
+	}
+
+	newCID, err := addDirToCluster(filepath.Join(tmpDir, ce.config.MFSDir), ce.config.PinName, clusterClient)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := copyFilesToMFSDir(ctx, *newCID, ce.config.MFSDir, sh); err != nil {
+		return nil, err
+	}
+
+	// If there's a stale cache pinned in the IPFS cluster, update the pin with the new CID and unpin the previous one
+	if prevPinCID != nil && prevPinCID.Cid != newCID.Cid {
+		if err := updatePin(ctx, ce.config.PinName, prevPinCID, newCID, clusterClient); err != nil {
+			return nil, err
+		}
+
+		logrus.Debugf("unpinning previous pin: %s\n", prevPinCID.Cid)
+		_, err = clusterClient.Unpin(ctx, *prevPinCID)
```

Comment:
Is it enough to only unpinning 1 CID? Isn't it possible that previously multiple CIDs being associated to that pin name?